rmf_task_ros2/src/rmf_task_ros2/bidding/Auctioneer.cpp

```diff
@@ -0,0 +1,307 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <rmf_task_ros2/bidding/Bidding.hpp>
+#include <rmf_task_ros2/bidding/Auctioneer.hpp>
+
+namespace rmf_task_ros2 {
+namespace bidding {
+
+//==============================================================================
+class Auctioneer::Implementation
+{
+public:
+  std::shared_ptr<rclcpp::Node> node;
+  rclcpp::TimerBase::SharedPtr timer;
+  BiddingResultCallback bidding_result_callback;
+  std::shared_ptr<Evaluator> winner_evaluator;
+
+  struct BiddingTask
+  {
+    BidNotice bid_notice;
+    builtin_interfaces::msg::Time start_time;
+    std::vector<bidding::Submission> submissions;
+  };
+
+  // non-sequential tasks
+  std::map<TaskID, BiddingTask> ongoing_bidding_tasks;
+
+  // sequential bidding tasks
+  bool sequential;
+  bool bidding_in_proccess = false;
+  std::queue<BiddingTask> queue_bidding_tasks;
+
+  using BidNoticePub = rclcpp::Publisher<BidNotice>;
+  BidNoticePub::SharedPtr bid_notice_pub;
+
+  using BidProposalSub = rclcpp::Subscription<BidProposal>;
+  BidProposalSub::SharedPtr bid_proposal_sub;
+
+  Implementation(const std::shared_ptr<rclcpp::Node>& node_,
+    const bool sequential)
+  : node(node_), sequential(sequential)
+  {
+    // default evaluator
+    winner_evaluator = std::shared_ptr<LeastFleetDiffCostEvaluator>(
+      new LeastFleetDiffCostEvaluator());
+
+    const auto dispatch_qos = rclcpp::ServicesQoS().reliable();
+
+    bid_notice_pub = node->create_publisher<BidNotice>(
+      rmf_task_ros2::BidNoticeTopicName, dispatch_qos);
+
+    bid_proposal_sub = node->create_subscription<BidProposal>(
+      rmf_task_ros2::BidProposalTopicName, dispatch_qos,
+      [&](const BidProposal::UniquePtr msg)
+      {
+        this->receive_proposal(*msg);
+      });
+
+    timer = node->create_wall_timer(std::chrono::milliseconds(1000), [&]()
+        {
+          this->check_bidding_process();
+        });
+  }
+
+  /// Start a bidding process
+  void start_bidding(const BidNotice& bid_notice)
+  {
+    std::cout << "\n[Auctioneer] Add Bidding task_id: "
+              << bid_notice.task_profile.task_id << " to queue"<< std::endl;
+
+    BiddingTask bidding_task;
+    bidding_task.bid_notice = bid_notice;
+    bidding_task.start_time = node->now();
+
+    if (sequential)
+    {
+      queue_bidding_tasks.push(bidding_task);
+    }
+    else
+    {
+      ongoing_bidding_tasks[bid_notice.task_profile.task_id] = bidding_task;
+      bid_notice_pub->publish(bid_notice);
+    }
+  }
+
+  // Receive proposal and evaluate
+  void receive_proposal(const BidProposal& msg)
+  {
+    auto id_ = msg.task_profile.task_id;
+    std::cout << "[Auctioneer] Receive proposal from task_id: "
+              << id_ << " | from: " << msg.fleet_name << std::endl;
+
+    // check if bidding task is "mine", if found
+    // add submited proposal to the current bidding tasks list
+    if (sequential)
```

Comment:
Parallel bidding will result in a faster bidding process. Yet when there are multiple pending bidding tasks, this will create an issue in reassignment on the fleet adapter side, thus not being used right now. If so should I remove this option for now?

Reply:
I think if we're not currently using or testing it, it's best to just leave it out of the API. It's preferable to add new features later one we know they work rather than need to remove features that turn out to be defective.